_posts/2016-01-22-pso.md

```diff
@@ -0,0 +1,136 @@
+
+### Static init
+For simple applications the PSO initialization format is often known in advance. To save the user from a burden of instantiating the `Init` struct, we provide a convenient extended macro to specify the values in place:
+{% highlight Rust %}
+gfx_pipeline!( pipe {
+    vertex: gfx::VertexBuffer<Vertex> = (),
+    const_locals: gfx::ConstantBuffer<Local> = "Locals",
+    tex_diffuse: gfx::TextureSampler<Rgba8> = "Diffuse"
+    buf_noise: gfx::ShaderResource<[i32; 4]> = "Noise",
+    pixel_color: gfx::RenderTarget<Rgba8> = "Color",
+    depth: gfx::DepthTarget<Depth> = gfx::state::Depth {
+        fun: gfx::state::Comparison::LessEqual,
+        write: false,
+    },
+});
+{% endhighlight %}
+This extended version will also generate the trio of structs, but in addition have the `Init::new()` method to give you an instance right away! Here is an example code for PSO creation after this macro is invoked:
+{% highlight Rust %}
+    let pso = factory.create_pipeline_state( // defined in `gfx::traits::FactoryExt`
+        &shaders, // `gfx::ShaderSet<R>` - has all the shaders
+        gfx::Primitive::PointList, // primitive topology
+        gfx::state::Rasterizer::new_fill(gfx::state::CullFace::Nothing),
+        Init::new() // our new shiny initializer
+    ).unwrap();
+{% endhighlight %}
+There is a simpler version of this function too:
+{% highlight Rust %}
+    let pso = factory.create_pipeline_simple( // defined in `gfx::traits::FactoryExt`
+        &vertex_shader, &fragment_shader,
+        gfx::state::CullFace::Nothing, Init::new()
+    ).unwrap();
+{% endhighlight %}
+
+### Drawing
+The resulting type will be `gfx::PipelineState<pipe::Meta>`, but fortunately the compiler can infer that for you. Using this PSO is rather trivial - the user is only required to construct the `Data` portion themselves:
+{% highlight Rust %}
+    let data = pipe::Data {
+        vertex: .., const_locals: ..,
+        tex_diffuse: .., buf_noise: ..,
+        pixel_color: .., depth: ..,
+    };
+    let slice = gfx::mesh::Slice {...};
+    encoder.draw(&slice, &pso, &data);
+{% endhighlight %}
+This is rather minimalistic, but, more importantly, there is next to no opportunity to shoot yourself in the foot! All the inputs of the PSO are guaranteed at compile time (if the type system is not abused). Performance-wise the solution is also perfect - all the mapping from the user data and the PSO inputs is already ready for us (contained in the `Meta`, which is constructed during PSO creation), so the actual binding doesn't involve any lookups.
```

Comment:
What sort of abuse do you have in mind, for "if the type system is not abused"? It seems like an unnecessary comment